File: cpp/141.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <deque>
#include <unordered_map>
#include <unordered_set>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
  bool hasCycle(ListNode *head) {
    unordered_set<ListNode*> saw;
    while(head){
      if(saw.find(head) != saw.end())
        return true;
      saw.insert(head);
      head = head->next;
    }
    return false;
  }
};

int main(){
  auto sol = Solution();
}

File: python/141.py
class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slow = fast = head
        while slow and fast:
            fast = fast.next
            if fast is slow:
                return True
            if fast:
                fast = fast.next
                if fast is slow:
                    return True
            slow = slow.next
        return False
